### PR TITLE
Get duplicates by ml

### PR DESCRIPTION
**Pull request by erezh31**

### CODE REVIEW

**Comment by meirwah** [comment]
@erezh31 some issue with your automation:
image

**Comment by meirwah** [comment]
@yaakovi can you review too?

File: Scripts/script-GetDuplicatesMl.yml [hunk 1]
```diff
@@ -0,0 +1,641 @@
...
+          return labels_map
+
+      @staticmethod
+      def get_domains(indicators, labels):
+          domains = []
+          for email in indicators.get('Email', []):
+              domains.append(Utils.extract_email_domain(email))
+
+          for url in indicators.get('URL', []):
+              domains.append(Utils.extract_domain_from_url(url))
```
Comment (meirwah) [inline Scripts/script-GetDuplicatesMl.yml]: there might be indicator of type 'Domain', you need to user that too, generated by:
https://github.com/demisto/content/blob/master/Scripts/script-ExtractDomain.yml

File: Scripts/script-GetDuplicatesMl.yml [hunk 2]
```diff
@@ -0,0 +1,641 @@
...
+              return {}
+
+          email_user_names1 = map(Utils.extract_email_username, emails1)
+          email_user_names2 = map(Utils.extract_email_username, emails2)
+
+          features = Utils.calculate_similar_pattern('email_username', email_user_names1, email_user_names2)
+
+          features['email_min_ld'] = Utils.get_minimal_edit_distance(email_user_names1, email_user_names2)
+
+          return features
```
Comment (meirwah) [inline Scripts/script-GetDuplicatesMl.yml]: Try to look at numpy maybe for those array manipulation

File: Scripts/script-GetDuplicatesMl.yml [hunk 3]
```diff
@@ -0,0 +1,641 @@
...
...9\xea\xfa+*\xff\xb6~,\xbfl\x1e\x923q\xd5\xfb\xf4\xff\\\xdc\xf2R\x8c\xa0\x1d$\x85\xec@\x0f\x16d9\xbe\xfe\xf9V=_\xfa\xef\xc7\x83?N!=\x9f\x9d\xb2q\t\x95/\xef<\xaa,{\xa1\xfa#\xfa\x9d\x0eov\xf4!*\xda\x14\xad\xfe\xe3\xf6\x02\xd4\xf6\x81\xfb\xc4^\xa7\xde\xa0k\x1b/\xa5\x9ai\xb7\xa4nF&\xde\x88\x7f\xde\x92:\x13\xdde\xfa\xf2\xddv\xd4\xba!V\xb1\x87v\xd2\x94k\x17\x8f\xa2\x93\xe2>\x94\xe3\xc8\xc5\xad\xf6\\\xefK9\xf8D7O\xbd\xe2LMou!$|F\x7f\xaav\xd7\x16\x92\x94{.TZ\xbbj^\xe1\x85\x03(\xab:\x9fvE\xf6\x18DY\xbd,\\\x90\xd6j0ub\xecz\xd3\xf8k#\xa9\xc6\x0f\xe8"W\x8d\xb1T\xfb\xa5\x9f\x12\xe2\x1b\x8c\xa5\x96{\xa0\xa4\xc8\xd9\x13\xa8V7:~\xe9\xfd\xc5\x9f\xa2\x1c^\xe55\xd6\x0c\xa0Pv\xd1\xed\x03}\x03\xa8K\r3\x1en\xfd\x9d\xa1v51/\xaek\x1cL\xdd\xcbL7\x0e?\x1dL\r\xd2n&\xb90"\x942\xfe8\xcaJk\xf6\x14j\xe5\xd9\xcf}\'<\x99Ie\xb7\xdev1H}\x16\x15r\xab{_\xef\xcc\x08*y\xfc\xaf\xc3\x1c\xb2"(\xf3\xdf\xfa\xd9v\x90R\xb7F\xcd!c\x1fEP\xa7\x12\x9f\xa6\xd8\xfe1\x8f2\x9c}\xaf\x81fX$e\xdd\xe5\xd9\xaa\xabR\xaa\xed\x1e\xb0\xa1{D$52\xabm\xdc\xe1CQ\xd4\x88q\xe7\x9f\x1ci\xbe\x88jU}\xce\xea^_\x17S\xcd\x9a\x16\r\xb8\x9c\xb9\x8cj/)\xda\xbc\xf5\xcf\x95\xd4\xb2-^5B\xfb\xaf\xa6\x82\xdc\xedZ\xfa\x05\xad\xa5\x06<X\xedn\x92\xb5\x99\xda\x90\x90i
```
Comment (meirwah) [inline Scripts/script-GetDuplicatesMl.yml]: you need to impl this

Comment (meirwah) [inline Scripts/script-GetDuplicatesMl.yml]: *Duplicate incidents candidates*
sounds more accurate to me

File: Scripts/script-GetDuplicatesMl.yml (hunk 1 above)
Comment (erezh31) [reply Scripts/script-GetDuplicatesMl.yml]: handle this

File: Scripts/script-GetDuplicatesMl.yml (hunk 2 above)
Comment (erezh31) [reply Scripts/script-GetDuplicatesMl.yml]: V

File: Scripts/script-GetDuplicatesMl.yml (hunk 3 above)
Comment (erezh31) [reply Scripts/script-GetDuplicatesMl.yml]: ok

Comment (erezh31) [reply Scripts/script-GetDuplicatesMl.yml]: ok

**Comment by erezh31** [comment]
@mayairv please review

File: Scripts/script-GetDuplicatesMl.yml (hunk 3 above)
Comment (erezh31) [reply Scripts/script-GetDuplicatesMl.yml]: V

Comment (erezh31) [reply Scripts/script-GetDuplicatesMl.yml]: V

File: Scripts/script-GetDuplicatesMl.yml (hunk 2 above)
Comment (erezh31) [reply Scripts/script-GetDuplicatesMl.yml]: V

File: Scripts/script-GetDuplicatesMl.yml (hunk 1 above)
Comment (erezh31) [reply Scripts/script-GetDuplicatesMl.yml]: V

File: Scripts/script-GetDuplicatesMl.yml [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
...e6\xb4I=\xbaW\xd3\xc2\xf1_S\xea\xd1^\xb3\xd3\xc3\x9b\xd7\xd6\xa5\xcd\xeb\xd5zsMW\x97n\x15Z\xed\xee\xa6\x06\xba\xf4\x91W\x9fS{\x19\xe8\xd2\xb1\xe3g\xa7/\xb5i@w\xbc\xd2\xadn\x90]\x03\xfaq\xec\x9f\x89\xb3g4\xa0k.l\xbemP\\\x03z@\xc3E\x0b\'\x9b\x1a\xd0\xcfF;\x9e\xda\xf6\xd8\x88^?uC\x8e\xd7\xf0F\xb4^\xc7\xee\x9fo\x8emDw\x1fx]\xef\x8e\x94\xd7M\xef\x1d{5\xbd1=e\x90\xb1\x81dNc\xfa(57\xa7\xc1\x8e\xc6t\xb5\xf5\xda\xf7\xec?7\xa6GEv6\x18\xd4\xb5\t]<\xffB\xaa\xe1\xa4&\xf4\x94\xac\rC\xee\xcfnB\xdfP\xcb8\x93X\xcf\x98\xbe6\xe1\xd6\xc4$\x17c\xfa\xf2\xb3\x0b\x89\x07]\x8d\xe9\xfcE\xe6\xfe\xdb\x06\x18\xd3\xb5\xab\xb5\xed+\x19lL\x1f5\x9e_\xd8k\xa91}v\xa9\xdd\x9d\xc7[[\x94\x94\x7f\x91\xbe\xad\x05=6\xce\xf9r\xf6\xf9\x16t\xc06\xf5\x13\xce\x05-\xe8/>\xef4w\xa9\x9b\xd2\xa6~\xd3\xe2B\xeb\x98\xd2\x17\xdb\'\xbd\xb8/ez\x8f\xe9\x9d\xbb\xd75\xa5m-w\x1f4\xd15\xa5c\x06O\x0e\xeb\xdd\xcc\x94\x16\x1f|\xb2\xd2o\x83)}\xd9\xdd\xa5`\xcbFS\xda!"6\xab\xe6fS\xfa]\x9egB\xfaVS\xfac\xd7\xa0\xabz\xd1\xa6\xb4t{h~\xfd\x9d)\x1d]\xd0o\t\xf5\xde\x94>\x99\xab\x7f2A\xca\xc0\xe2\xf9\xce\xeb\xc5ft\xe2\x1b\xf3\xa6+\x83\xcc\xe8F9#\xd7\x0e\x98iF\xb7;\xf0\x9b\xcd\x81EftJ\xef\xb4\xc1\x9f\xee\x9a\xd1.\xa3"\x8d&5oE\xfb\xe7\x7f\xbcp\xa1\x8f\x94\x
```
Comment (mayairv) [inline Scripts/script-GetDuplicatesMl.yml]: Consider duplicate incident examples from the local environment within the last "X" days. Default is 0, which means do not use.

Comment (mayairv) [inline Scripts/script-GetDuplicatesMl.yml]: Consider the equity of the IP base on this subnet mask (a valid value is a number between 0 - 32). Default is 32 which means an exact match of the IP addresses.

File: Scripts/script-GetDuplicatesMl.yml [hunk 5]
```diff
@@ -79,7 +83,8 @@ script: |
           index = email.rfind("@")
           if index == -1:
               return None
-          return email[index + 1:]
+          return Utils.extract_domain_from_url(email[index + 1:])
```
Comment (meirwah) [inline Scripts/script-GetDuplicatesMl.yml]: can't you have here out of bounds? in case @ is in the end?

Comment (erezh31) [reply Scripts/script-GetDuplicatesMl.yml]: No, python slice are safe. it will return just empty string

**Comment by meirwah** [comment]
@erezh31 let's stable and merge his branch

**Comment by meirwah** [comment]
image

@erezh31 plz look to fix this

**Pull request merged**